### ISSUE TITLE
Create and configure elements in child-before-parent order

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#481]
For instance `<option>` elements for a  `<select>` need to be added before the `value` attribute for a select is added

**Comment by mstahv** [comment]
Decision makers, please also see the discussion in the closed #3651 to see the importance of this issue. If this wont be solved soon properly, we need a lot of quality documentation to play down the WTFs this causes.

**Comment by amahdy** [comment]
Oh hey there, my bug was closed so it seems I need to bug this one. Initial impression feels that there will be a *lot* of complains about this problem. I was told that few people internally run into this as well.

**Comment by mstahv** [comment]
I'm trying to build a flow wrapper for this component: https://github.com/miztroh/wysiwyg-e/blob/v2.1.0/README.md

I guess it is this bug that makes it impossible to configure e.g. available heading levels from Flow. The web component reads them when the "wysiwyg-tool-heading" element is attached, but Flow sets them too late? Only way to configure this component would be to use template?

**Comment by mstahv** [comment]
It might be that my issue with the wysiwyg-e is not about this. Don't know for sure. With Flow a menu is not visible, without it displays fine.

**Comment by Legioth** [comment]
In general, it's a bug in the web component if it doesn't support incremental updates (i.e. it doesn't react to changes after it has been attached).

Still, it's quite common that web component implementations have such bugs, so this is not a reason for Flow not to do things in an order that would work better.

**Comment by samie** [comment]
Yes, I ran into this one also. If not supported, we need to document how to make properly Vaadin Flow compatible web components. By default web components are not really designed in a dynamic way and are primarily meant to be used in a rather static HTML document not in a SPA.

**Comment by samie** [comment]
I very much agree what @amahdy already said in #3651 ("The logic of execution is different from what I wrote in Java, add(..) is the last line of the code so I expect it to happen last."). Would also expect to keep the order of method calls:

        HtmlContainer parent = new HtmlContainer("parent-element");
        HtmlContainer child = new HtmlContainer("child-element");
        child.getElement().setAttribute("attr", "value"); // attr value used in component ready
        parent.add(child); // builds "detached" parent-child tree
        add(parent); // attaches everything once, ready invoked

i.e at least in this case I'm not sure, if it is really needed to be "child-before-parent", but just _maintaining the order_ of element API  calls.

**Comment by mstahv** [comment]
I think it should.

**Comment by samie** [comment]
Yes, I think sounds logical and generally understandable. I would expect symmetric behaviour (as much as possible) of elements/components both on client and server side.

**Comment by mstahv** [comment]
We have absolutely no idea how *focus()* works from the frameworks perspective. I think it is up to the component implementation to fix the functionality in that case. The Java API (and possibly some framework helpers) or the web component should fix the broken web.

**Comment by Legioth** [comment]
The framework's own `focus()` implementation could be made to work in whatever way is necessary. I'm referring to any custom JS that directly or indirectly calls something like `focus()` or `document.getElementById("foo")` that won't work before the DOM tree has been constructed and attached.

The hypothesis is that this ordering would work in 99% of all cases, and for the remaining cases it would have to be possible to do something to override the ordering.
1. Create element instance
2. Attach all fully configured children
3. Set all properties and attributes
4. Attach the element to its parent (recursively, thus attaching it to the DOM)
5. Call any JS

The benefits of having a hardcoded ordering would be:
1. Multiple assignments to the same property or attribute can be coalesced to avoid wasting network bytes and client-side CPU on things that will be overwritten "immediately".
2. Running JS only after things have been set up. This is analogous to the way V7 first updates all shared state, then fires state change events for everything and only after that running RPC.
3. Setting up as much as possible before attaching the component (triggering e.g. Polymer's `ready` callback).

**Comment by samie** [comment]
Yes, sounds logical on "component level" APIs. On the other hand, on "DOM level" it would make sense to keep it as close as possible to original DOM logic. Then you can at least predict when some method call might fail because of detached state and change the order in your own code.

**Comment by Legioth** [comment]
The problem is that component and element "levels" are intermingled because the application developer might call component methods in an arbitrary order, which would in turn cause element methods to happen according to the same order. The ordering would thus only be effective for things that happen within one component-level method call.

**Comment by Legioth** [comment]
There are basically two different classes of this problem.
* The simple case is when there's only some specific property or attribute that absolutely needs to be set before attaching. An example of this is the official `<google-map>` web component where `apiKey` must be set from the beginning, and you cannot even change it afterwards without causing problems.
* The difficult case is when the entire web component is unresponsive to changes after attaching. For these cases, changing the ordering of operations would still only help with initialization, whereas incremental updates originating from user interaction would still not work.

It might be possible to do a workaround for the simple case that wouldn't require any heavy refactoring. The idea would be to add a low-level feature for assigning  a "post create" (or "pre attach") JS snippet to a newly created `Element` instance. This would thus completely bypass the regular logic for setting children, properties and attributes - those would still happen after attaching in the same way as now.

**Comment by amahdy** [comment]
The second use case is not our responsibility. If the webcompoenent is designed not to respond to changes why would we want it to respond?

It feels that you want to solve problems for a forecasted bad code, like bad WC design and `setFocus` when component is not attached yet. While in my opinion this is not our responsibility. Or at least the most important to have is: execute code in the correct order. As a developer I expect the logic I write to go as intended and not be part of a special case as mentioned in the workaround.

**Comment by amahdy** [comment]
Quick discussion with @samie , here is what I expect:

```js
let x = document.createElement("div");
x.setAttribute("foo", "bar");
document.body.appendChild(x);
```

Result is:

```html
<body>
 <!-- .. -->
 <div foo=​"bar">​</div>​
</body>
```

Straightforward. This is what any developer would expect. `appendChild` should go to the client and append whatever I have created before that line. All other use cases / error preventions are just side cases not the norm.

Also if a developer write a wrong code and after testing it finds that it works, he will be puzzled on *why it works?!*. We don't normally expect framework to correct our semantic errors without explicitly telling me so.

**Comment by Legioth** [comment]
The final result is typically the same regardless of reordering or coalescing things. The interesting question is which intermediate states there will be.

One rather extreme example is a sequence like this:
```
parent.appendChild(Element.createText(oneMegabyteOfText));
parent.removeAllChildren();
```
Should this lead to sending a 1mb message to the browser, or should the framework be "smart" and leave out things that won't be included in the end result?

**Comment by amahdy** [comment]
My use case shows that it is not the same. I like to describe it as a constructor for the webcomponent. Attributes during initialize time are constructors that accept parameters. Initializing a component without any possibility to have attributes is equal to no possibility to have a non-empty constructor.

If I understand your examples correctly, you are mainly trying to lower the number of round trips to the client side? While this sound nice to have but again it’s a bad code of the developer so this shouldn’t be the priority. The priority is to execute the code as intended.

How is the situation now? Do you render the component on the server side and evaluate? Doesn’t seem so because for some reason the client side initialize an empty component first regardless of all Boolean operations your perform on the server side.

**Comment by willtemperley** [comment]
I've been working on a Flow wrapper for the LeafletJS mapping library where I've been running into this issue. 

Essentially Flow doesn't support overloaded constructors for web components. Take the example of a web component wrapping a JS object with two optional constructor arguments [1].  With the current design I would have two observe two properties and delay construction until I get all the arguments. The problem is I don't know how many arguments the server-side code is going to send, so the problem is intractable. 

The workaround / hack I've implemented instead is declaring a "constructorargs" string property on all components, which is actually a JSON serialisation of the constructor arguments.  As I know the number of arguments, I can easily call the correct factory method to construct an object.  It's messy but it works.

Unfortunately I think that unless Flow starts to support the same semantics as Polymer and web components in general, there will be a plethora of Vaadin-specific components created (as I'm doing) and a significant proportion of third-party web components will be unusable in Vaadin.

[1] http://leafletjs.com/reference-1.3.0.html#popup

**Comment by Legioth** [comment]
> The "ready" event is fired directly after you run `document.createElement` so there is no time to attach the children before the "ready" event

This is no longer true as of Polymer 2. This change is based on changes in the custom element specification that restricts what can be done in a custom element constructor. It seems like nowadays, `ready` is triggered the first time `connectedCallback` is run (but still always before `conenctedCallback` triggers Polymer's own `attached` callback).

**Comment by Legioth** [comment]
I realized that we can actually treat creation and initial binding for an element as a special case and in that way fix anything related to `ready`, `attached` and `connectedCallback` with a quite easy fix, which is in https://github.com/vaadin/flow/pull/4146.

This means that for a newly created element, the ordering will always be:
1. Add `$server` and `@EventHandler` stubs
2. Attach children, virtual children and any server-provided shadow root
3. Set inline styles and class names
4. Set attributes and properties
5. Add DOM listeners that do RPC to the server
6. Attach the element to its parent

Incremental updates to an existing element will still happen in an arbitrary order.

**Comment by Legioth** [comment]
I'll consider this issue closed since the order of events that affect `ready` and `

**Comment by Legioth** [comment]
I'll consider this issue closed since the main issue about components that read initial settings in `ready`, `attached` or `connectedCallback` has been fixed.

I've created https://github.com/vaadin/flow/issues/4157 for also applying incremental updates for an existing element in the same order.

Please create a separate issue if there's something else that has been discussed here that you still feel would be relevant to do something about.
